### ISSUE TITLE
Detect circular object structures in JSON.stringify

### ISSUE THREAD

**Issue by echenley** [Microsoft/TypeScript#29415]
## Search Terms

- "JSON stringify"
- "TypeError: Converting circular structure to JSON"
- Various iterations of the above

## Suggestion

The [JSON.stringify built-in type](https://github.com/Microsoft/TypeScript/blob/master/lib/lib.es5.d.ts#L1056) does not currently protect against circular object structures. For instance, the following runtime error is not caught by TS: 

```ts
const obj: any = {};
obj.self = obj;

type Circular = {
  self: Circular
}

const typedObj: Circular = obj;

// Can TS determine that this will throw?
JSON.stringify(typedObj);
// TypeError: Converting circular structure to JSON
```

I am skeptical that this is even possible. At the very least, it seems difficult to express.

## Use Cases

Preventing `JSON.stringify` bugs at compile-time rather than runtime.

## Examples

http://www.typescriptlang.org/play/#src=const%20obj%3A%20any%20%3D%20%7B%7D%0D%0Aobj.self%20%3D%20obj%3B%0D%0A%0D%0Atype%20Circular%20%3D%20%7B%0D%0A%20%20%20%20self%3A%20Circular%0D%0A%7D%0D%0A%0D%0Aconst%20typedObj%3A%20Circular%20%3D%20obj%3B%0D%0A%0D%0A%2F%2F%20Can%20TS%20determine%20that%20this%20will%20throw%3F%0D%0AJSON.stringify(typedObj)%3B%0D%0A%2F%2F%20TypeError%3A%20Converting%20circular%20structure%20to%20JSON

## Checklist

My suggestion meets these guidelines:

* [ ] This wouldn't be a breaking change in existing TypeScript/JavaScript code
  * *This would be breaking in that it could cause new compiler errors.*
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by RyanCavanaugh** [comment]
This seems 100% doomed - what you call `Circular` I call `LinkedList` or `BinaryTreeNode`, which are totally legal to stringify when they don't contain cycles.

What you need here is some notion of tracking *object identity*, then another notion of tracking the current *deep* data content of a variable, neither of which TS currently does.

**Comment by echenley** [comment]
The `Circular` type above is not a linked list or binary tree node because it never terminates. If the type were written as follows, I would agree:
 
```js
type Circular = {
  self?: Circular
}
```

But assuming you are using `strictNullChecks`, then the following type represents either a linked list with infinite nodes, or a circular object, both of which are problematic for `JSON.stringify`:

```js
type Circular = {
  self: Circular
}
```

Like I said, I'm not sure it's possible to write a type that disallows this kind of construct. Still, it's worth documenting that this is a class of bugs that cannot be prevented by TS.
